File: 2020/14/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', $ARGV[0] || 'input.txt'));

my @memory;
my $mask_set = 0;
my $mask_clr = 0;

while (my $line = <$fh>)
{
	if ($line =~ /^mask = ([01X]{36})/)
	{
		my $mask_val = $1;
		$mask_set = 0;
		$mask_clr = 0;
		for (my $i = 0; $i < 36; $i++)
		{
			my $c = substr($mask_val, $i, 1);
			$mask_set |= (1 << (35 - $i)) if ($c eq '1');
			$mask_clr |= (1 << (35 - $i)) if ($c eq '0');
		}
	}
	elsif ($line =~ /^mem\[(\d+)\] = (\d+)/)
	{
		$memory[$1] = ($2 & ~$mask_clr) | $mask_set;
	}
	else
	{
		die "* Bad line\n"
	}
}

my $answer1 = 0;
foreach my $val (@memory)
{
	$answer1 += $val;
}

printf("Part 1 answer: %d\n", $answer1);

File: 2020/14/both.py
#!/usr/bin/python

import string
import sys
import re

def masked_inc(val, mask):
	incval = (val | ~mask) + 1
	return (val & ~mask) | (incval & mask)

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r');

mask = re.compile(r"^mask = ([01X]{36})")
mem = re.compile(r"^mem\[(\d+)\] = (\d+)")

memory = {}
memory2 = {}
mask_set = 0
mask_clr = 0
mask_float = 0
float_cnt = 0

for line in fh:
	m = mask.match(line)
	if m:
		mask_val = m.group(1)
		mask_set = 0
		mask_clr = 0
		mask_float = 0
		float_cnt = 0
		for i in range(36):
			if mask_val[i] == '1':
				mask_set |= (1 << (35 - i))
			elif mask_val[i] == '0':
				mask_clr |= (1 << (35 - i))
			else:
				mask_float |= (1 << (35 - i))
				float_cnt += 1
	m = mem.match(line)
	if m:
		addr = int(m.group(1))
		value = int(m.group(2))
		memory[addr] = (value & ~mask_clr) | mask_set
		addr2 = (addr | mask_set) & ~mask_float
		for n in range(1 << float_cnt):
			memory2[addr2] = value
			addr2 = masked_inc(addr2, mask_float)

answer1 = sum(memory.values())
answer2 = sum(memory2.values())

print("Part 1 answer: %d" % answer1)
print("Part 2 answer: %d" % answer2)
